Use phpstan/phpdoc-parser to retrieve additional type information from PhpDoc

Namoshek opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Doc updated   | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1259
| License       | MIT

As suggested by @goetas in #1259, this PR fixes the mentioned issue by implementing `phpstan/phpdoc-parser` support to read `@var` tags from PhpDoc. Only one test case for single-line PhpDoc parsing has been added, which works fine now.

In my opinion, a BC break is introduced by adding a constructor to `DocBlockTypeResolver`, since none existed before (which means possible extensions probably do not contain and call to the parent constructor so far).

This needs to be reviewed carefully, as I'm not sure if I considered all possible cases. The tests seem to work fine though.

goetas commented on composer.json:
```diff
@@ -35,6 +35,7 @@
         "doctrine/phpcr-odm": "^1.3|^2.0",
         "jackalope/jackalope-doctrine-dbal": "^1.1.5",
         "ocramius/proxy-manager": "^1.0|^2.0",
+        "phpstan/phpdoc-parser": "^0.4",
```
i think that this should stay in the `require` section, not `require-dev`

Namoshek replied:
You are totally right. No idea how it ended up in the dev dependencies.

goetas commented:
except only one comment, this looks good to me. @dgafka can you also check it please?

goetas commented:
since https://github.com/schmittjoh/serializer/pull/1256 has been merged, this needs to be rebased.

Namoshek commented:
I changed the dependency location and rebased.

#1256 is very welcomed by me as well by the way. since I was about to write a custom handler for `CarbonInterface` and realized it doesn't work just yet. 😄

goetas commented:
yay!

goetas commented:
thanks a lot!
